KAI - 텍스트 추가하기

joho2022 opened the pull request:
## 🎯주요 작업

- [x]  settingsPanelViewController에서 직접 notification을 받도록 개선하기
- [x]  VisualComponent 자체를 key로 쓸 수 있는 방법도 생각하고 적용하기
- [x]  드래그할 때 투명도 안생기는 버그 고치기
- [x]  텍스트 추가하기
    - [x]  랜덤한 위치 지정해서 위치부터 공백기준 5단어 표시하기

## 📚학습 키워드

- 내부속성을 쓰는 것보다 객체 자체를 Key로 사용하는 것이 좋음. 유니크아이디는 같아도 속성이 다를 수 있기 때문에~

## 💻고민과 해결

## 💁‍♂️`selectedView.frame.origin.x`는 하위까지 다 풀어쓰고 `newWidth`는 할당했는 데 일관성있게 하셔도 좋을 것 같네요

### 🧙‍♂️나의 해결

```swift
// main뷰컨트롤러 handlePointUpdate메서드
let newWidth = newSize.width
let newheight = newSize.height
let selectedViewOriginX = selectedView.frame.origin.x
let selectedViewOriginY = selectedView.frame.origin.y
        
selectedView.frame = CGRect(x: selectedViewOriginX, y: selectedViewOriginY, width: newWidth, height: newheight)
```

## 💁‍♂️이 부분도 settingsPanelViewController에서 직접 notification을 받을 수 있지 않을까요?

```swift
// main뷰컨트롤러 handleColorChanged메서드 
self.logger.info("배경색 변경 수신완료!")
updateViewBackgroundColor(for: rectangleView, using: randomColor)       
self.settingsPanelViewController.backgroundStack.updateColorButtonTitle(with: randomColor)
```

### 🧙‍♂️나의 해결

```swift
// 변경될 때 rectangleColorChanged 수신받으면 업데이트 하도록 함 
// main과 setting 둘다 .rectangleColorChanged 수신받음
@objc private func handleColorChanged(notification: Notification) {
        guard let randomColor = notification.userInfo?["randomColor"] as? RGBColor else { return }
        
        self.logger.info("배경색 변경 수신완료(Setting)!")
        backgroundStack.updateColorButtonTitle(with: randomColor)
    }
```

## 💁‍♂️uniqueID를 key로 쓸수도 있지만, VisualComponent 자체를 key로 쓸 수 있는 방법도 생각해보세요.

```swift
private func findView(for component: VisualComponent) -> UIView? {
        return viewRegistry[component.uniqueID]
```

### 🧙‍♂️나의 해결

현재 Swift의 타입 시스템에서는 프로토콜 자체에 직접 **`Hashable`**을 채택할 수 없으며, 이는 구조적 한계로 인해 프로토콜 타입은 **`Hashable`**을 직접 채택할 수 없다.

"제네릭 프로토콜"을 콜렉션 타입으로 설정하면 컴파일 에러 발생

```swift
// 타입소거를 사용하여 키값으로 사용하고자 함.
struct AnyVisualComponent: Hashable {
    private(set) var component: VisualComponent
    
    init(_ component: VisualComponent) {
        self.component = component
    }
    
    func hash(into hasher: inout Hasher) {
        hasher.combine(ObjectIdentifier(type(of: component)))
        hasher.combine(component.getUniqueID())
    }
    
    static func == (lhs: AnyVisualComponent, rhs: AnyVisualComponent) -> Bool {
        return lhs.component.getUniqueID() == rhs.component.getUniqueID()
    }
}
```

 **`VisualComponent`** 프로토콜을 준수하는 어떤 객체든지 **`viewRegistry`** 딕셔너리의 키로 사용할 수 있도록  **`AnyVisualComponent`**는 내부적으로 **`VisualComponent`** 프로토콜을 준수하는 객체를 감싸고, 이를 통해 **`Hashable`** 프로토콜 요구사항을 만족시키며, 딕셔너리의 키로 사용하여 해결하고자 함.

**[고민]**속성에 직접 접근하는 대신, 값을 가져오거나 설정하는 메서드를 프로토콜에 정의하는 방법을 생각했는데, 데이터관리를 할 때 사진모델과 사각형모델을 구분하지 않고 공통된 프로토콜을 채택하는 객체를 키값으로 데이터 관리하고 싶은데, 이때 get set구조를 하지 않고 객체지향 원칙을 지키면서 앱 돌아가는 구조에 대해 견문이 부족하여.. 방법이 떠오르지 않습니다.. 여러 방법을 모색했지만 결국 프로토콜에 get set구조가 선언되는 방법만 떠오릅니다.….💧😭😢😿🥲💦

## 💁‍♂️이 메소드는 settingsPanelViewController에 있어도 될 것 같은 내용이 많은 것 같습니다. 어떻게 생각해요?

```swift
// main뷰컨트롤러 viewTapped메서드 안에
        updateColorButtonTitle()
        updateStepperValueTitle()
    }
    
    private func updateColorButtonTitle() {
        if let selectedView = selectedView,
           let uniqueID = findKey(for: selectedView),
           let component = plane.findComponent(uniqueID: uniqueID) {
            if let color = component.backgroundColor {
                self.settingsPanelViewController.backgroundStack.updateColorButtonTitle(with: color)
            } else {
                self.settingsPanelViewController.backgroundStack.updateColorButtonTitle(with: nil)
            }
        } else {
            self.settingsPanelViewController.backgroundStack.updateColorButtonTitle(with: nil)
        }
    }
    
    private func updateStepperValueTitle() {
        if let selectedView = selectedView,
           let uniqueID = findKey(for: selectedView),
           let component = plane.findComponent(uniqueID: uniqueID) {
            let newX = Double(component.point.x)
            let newY = Double(component.point.y)
            let newWidth = Double(component.size.width)
            let newHeight = Double(component.size.height)
            
            self.settingsPanelViewController.pointStack.updateStepperValue(firstValue: newX, secondValue: newY)
            
            self.settingsPanelViewController.sizeStack.updateStepperValue(firstValue: newWidth, secondValue: newHeight)
        } else {
            self.settingsPanelViewController.pointStack.updateStepperValue(firstValue: 0, secondValue: 0)
            
            self.settingsPanelViewController.sizeStack.updateStepperValue(firstValue: 0, secondValue: 0)
        }
    }
```

### 🧙‍♂️나의 해결

```swift
// setting뷰컨트롤러 
func updateUIForSelectedComponent(_ component: VisualComponent?) {
        updateColorButtonTitle(with: component?.getColor())
        updateStepperValues(with: component)
    }

private func updateColorButtonTitle(with color: RGBColor?) {
        let color = color ?? nil
        backgroundStack.updateColorButtonTitle(with: color)
    }
    
private func updateStepperValues(with component: VisualComponent?) {
        let positionX = component?.getPoint().x ?? 0
        let positionY = component?.getPoint().y ?? 0
        let width = component?.getSize().width ?? 0
        let height = component?.getSize().height ?? 0
        
        pointStack.updateStepperValue(firstValue: Double(positionX), secondValue: Double(positionY))
        sizeStack.updateStepperValue(firstValue: Double(width), secondValue: Double(height))
    }
```

## 🤔결과
![텍스트추가하기](https://github.com/codesquad-members-2024/swift-drawing/assets/104732020/22345222-e5f8-4914-9e58-936f2e9f3394)

joho2022 committed:
README 업데이트: 텍스트 추가하기 내용

joho2022 committed:
feat: 텍스트 추가하기

godrm commented on DrawingApp/DrawingApp/LabelModel/Label.swift:
```diff
@@ -0,0 +1,71 @@
+        return backgroundColor
+    }
+    
+    func getUniqueID() -> UniqueID {
+        return uniqueID
+    }
```
getter / setter도 필요한 경우가 있고 아닌 경우가 있는데 그걸 고민해보면 됩니다
예를 들어 UniqueID 경우는 객체 단위로 비교하면 이걸 가져갈 필요가 없을 겁니다
그런 식으로 줄일 수 있는 부분을 줄여야 하는거구요. 뷰에 전달해야 하는 속성은 어쩔 수 없을 겁니다.

godrm commented on DrawingApp/DrawingApp/MainViewController.swift:
```diff
@@ -86,13 +90,15 @@ extension MainViewController {
+              let selectedModel = userInfo["selectedModel"] as? VisualComponent,
               let newSize = userInfo["size"] as? Size,
-              let selectedView = viewRegistry[uniqueID] else { return }
+        let selectedView = viewRegistry[AnyVisualComponent(selectedModel)] else { return }
```
질문에도 써있었는데 AnyVisualComponent로 감쌀 필요는 없고 VisualComponent가 AnyVisualComponent와 관련이 있게 만들면 됩니다. 그걸 고려해보세요

godrm commented on DrawingApp/DrawingApp/MainViewController.swift:
```diff
@@ -240,16 +256,30 @@ extension MainViewController: UIImagePickerControllerDelegate, UINavigationContr
+              let photoView = userInfo["photoView"] as? UIView else { return }
         
-        addPhotoViews(for: photoView, with: photoModel)
+        addViews(for: photoView, with: photoModel)
```
여기 private 메소드중에서 self.view 와 관련된 메소드들도 예를 들어 CanvasView 라는 커스텀 뷰를 만들고 거기에 메소드로 만들수도 있습니다. ViewController가 모든 것을 다 준비하고 self.view 에 지정하기 보다는 그게 더 간결할 겁니다

godrm commented on DrawingApp/DrawingApp/MainViewController.swift:
```diff
@@ -278,48 +308,50 @@ extension MainViewController: UIGestureRecognizerDelegate {
-                temporaryView?.frame = newFrame
-                
-                self.settingsPanelViewController.pointStack.updateStepperValue(firstValue: newX - (tempView.frame.size.width / 2), secondValue: newY - (tempView.frame.size.height / 2))
+                NotificationCenter.default.post(name: .componentPositionChanged, object: self, userInfo: [
+                    "tempView": tempView,
+                    "newX": newX - (tempView.frame.size.width / 2),
+                    "newY": newY - (tempView.frame.size.height / 2),
+                    "width": tempView.frame.size.width,
+                    "height": tempView.frame.size.height
```
이렇게 MainVC -> SettingVC 로 post 하는 방법도 있지만 
plane의 속성을 바꾸면 notification이 발생할테니 그걸 SettingVC에서 observe 해도 됩니다 
그 변화를 감지하는 게 더 좋습니다

godrm commented on DrawingApp/DrawingApp/Models/Plane.swift:
```diff
@@ -40,65 +42,93 @@ struct Plane: Updatable {
+        let width = data.size.width
+        let height = data.size.height
+        
+        let red = CGFloat(data.backgroundColor!.red) / 255.0
+        let green = CGFloat(data.backgroundColor!.green) / 255.0
+        let blue = CGFloat(data.backgroundColor!.blue) / 255.0
+        let opacity = CGFloat(data.opacity.rawValue) / 10.0
+        
+        let labelView = UILabel(frame: CGRect(x: pointX, y: pointY, width: width, height: height))
+        labelView.text = data.text
+        labelView.font = .systemFont(ofSize: 35, weight: .bold)
+        labelView.textColor = UIColor(red: red, green: green, blue: blue, alpha: opacity)
+        logger.info("텍스트 생성 명령하달!!")
+        NotificationCenter.default.post(name: .labelCreated, object: nil, userInfo: ["labelModel": data, "labelView": labelView])
```
이 부분 좀 이상한 것 같아요. Model 에서는 UIKit 에 있는 UILabel, UIColor 를 사용하지 않고 독립적으로 구현하셔야 합니다. UIView 생성하는 createRectangleView() 처럼 생성하는 것은 여기서 할 일이 아닙니다

godrm commented on DrawingApp/DrawingApp/PhotoModel/PhotoModel.swift:
```diff
@@ -23,6 +23,22 @@ class PhotoModel: VisualComponent {
+    
+    func getColor() -> RGBColor? {
+        return backgroundColor
+    }
```
PhotoModel은 backgroundColor가 필요없지 않던가요? 제외해봅시다
관련 메소드도 다 없어야겠죠. 그럴려면 Protocol을 분리해야 할 것 같네요

godrm commented on DrawingApp/DrawingApp/Protocol/AnyVisualComponent.swift:
```diff
@@ -0,0 +1,25 @@
+    
+    static func == (lhs: AnyVisualComponent, rhs: AnyVisualComponent) -> Bool {
+        return lhs.component.getUniqueID() == rhs.component.getUniqueID()
+    }
```
위에도 언급했지만 AnyVisualComponent 구체 타입말고 VisualComponent만 채택하고서도 Dictionary 키값으로 사용할 수 있는 방법을 고민해보세요. 이렇게 별도 타입을 구현할 필요는 없습니다

godrm reviewed (changes requested):
개선한 부분과 의도한 부분을 잘 이해했습니다
프로토콜 분리와 AnyVisualComponent, VC -> VC로 가는 Notification 은 개선해보고 넘어갑시다 
열어놓을테니 개선해보세요

joho2022 committed:
refactor: 피드백 개선

1. 프로토콜 분리와 AnyVisualComponent -> BaseRect 상속하게 하여 데이터 관리, RectColorful 이미지 구분하고자 함.
2. 드래그가 끝날 때 Palne 데이터 처리가 일어나고, 알림이 가서 드래그 중에는 임시뷰의 위치를 텍스트만 업데이트 하고자 함.
3. Plane에서 UIView를 생성하지 않고, UIView를 생성하는 ViewFactory 으로 개선

godrm commented:
개선한 커밋을 작성할 때도 커밋은 변화의 하나씩 남기세요. 커밋은 여러번 하고 push만 한 번에 하면 됩니다. 
특히 커밋 하나씩 적용하거나 되돌릴 수 있어야 하기 때문에 여러 변화를 포함하지 않도록 하세요

godrm commented on DrawingApp/DrawingApp/MainViewController.swift:
```diff
@@ -74,14 +73,14 @@ extension MainViewController {
     
```
다음에는 이 부분 Plane과 팩토리 구체 타입도 의존성을 역전하도록 외부에서 생성해서 넘겨보세요

godrm commented on DrawingApp/DrawingApp/MainViewController.swift:
```diff
@@ -148,7 +147,7 @@ extension MainViewController {
             }
             let selectedModel = findComponent(for: selectedRectangleView)!
-            plane.updateRectangleColor(uniqueID: selectedModel.getUniqueID())
+            plane.updateRectangleColor(uniqueID: selectedModel.uniqueID)
```
plane.updateRectangleColor()도 selectedModel을 그대로 넘길 수 있어도 될 것 같습니다.

godrm commented on DrawingApp/DrawingApp/Protocol/AnyVisualComponent.swift:
```diff
@@ -7,19 +7,19 @@
+//    static func == (lhs: AnyVisualComponent, rhs: AnyVisualComponent) -> Bool {
+//        return lhs.component.getUniqueID() == rhs.component.getUniqueID()
+//    }
+//}
```
사용하지 않는 코드는 과감하게 삭제하고 커밋하세요

godrm commented on DrawingApp/DrawingApp/Protocol/VisualComponent.swift:
```diff
@@ -8,22 +8,44 @@
+    }
+    
+    static func == (lhs: BaseRect, rhs: BaseRect) -> Bool {
+        return lhs.uniqueID == rhs.uniqueID
     }
     
     func hash(into hasher: inout Hasher) {
-        hasher.combine(getUniqueID())
+        hasher.combine(uniqueID)
     }
```
아마도 비교하는 hash를 처리하면 uniqueID는 private 으로 처리해도 될 겁니다.

godrm commented on DrawingApp/DrawingApp/Settings/SettingsPanelViewController.swift:
```diff
@@ -121,11 +104,11 @@ extension SettingsPanelViewController {
+        let positionX = component?.point.x ?? 0
+        let positionY = component?.point.y ?? 0
+        let width = component?.size.width ?? 0
+        let height = component?.size.height ?? 0
```
근데 어차피 nil이 넘어오는 경우가 없지 않을까요?

godrm commented on DrawingApp/DrawingApp/ViewFactory.swift:
```diff
@@ -0,0 +1,49 @@
+        labelView.font = .systemFont(ofSize: 35, weight: .bold)
+        labelView.textColor = UIColor(red: red, green: green, blue: blue, alpha: opacity)
+        
+        NotificationCenter.default.post(name: .labelCreated, object: nil, userInfo: ["labelModel": data, "labelView": labelView])
```
이 부분 post 이벤트가 Plane에서 이벤트와 동일한 것 같은데 이렇게 이벤트 전송하는 곳이 다른 곳에서도 보내면 혼란스러울 것 같습니다
왜냐하면 보통 post는 1곳에서 하고 받는 observer가 여러 곳인 1:n 구조가 많습니다.
그리고 엄밀히 View 생성과 Model 생성은 서로 다른 이벤트라고 봐야할 것 같습니다

godrm reviewed (approved):
수고하셨습니다. 다음 단계를 하면서 아래 내용을 개선해보세요
